Source/stash.cpp

```diff
@@ -0,0 +1,618 @@
+			Rectangle cell {
+				GetPanelPosition(UiPanels::Stash, StashCellPosition({ x, y })),
+				InventorySlotSizeInPixels
+			};
+
+			if (cell.Contains({ cursorPosition })) {
+				return { x, y };
+			}
+		}
+	}
+
+	return { -1, -1 };
+}
+
+void CheckStashPaste(Point cursorPosition)
+{
+	auto &player = Players[MyPlayerId];
+
+	SetICursor(player.HoldItem._iCurs + CURSOR_FIRSTITEM);
+	if (!IsHardwareCursor()) {
+		cursorPosition += Displacement(icursSize / 2);
+	}
+	Size itemSize { icursSize28 };
+
+	if ((itemSize.width & 1) == 0)
+		cursorPosition.x -= INV_SLOT_HALF_SIZE_PX;
+	if ((itemSize.height & 1) == 0)
+		cursorPosition.y -= INV_SLOT_HALF_SIZE_PX;
+
+	Point firstSlot = FindSlotUnderCursor(cursorPosition);
+	if (firstSlot == Point { -1, -1 })
+		return;
+
+	if (firstSlot.x + itemSize.width >= 10 || firstSlot.y + itemSize.height >= 10) {
+		return; // Item does not fit
+	}
+
+	// Check that no more then 1 item is replaced by the move
+	int16_t it = 0;
+	for (int y = 0; y < itemSize.height; y++) {
+		for (int x = 0; x < itemSize.width; x++) {
+			int16_t iv = abs(Stash.stashGrids[Stash.page][firstSlot.x + x][firstSlot.y + y]);
+			if (iv == 0 || it == iv)
+				continue;
+			if (it == 0) {
+				it = iv; // Found first item
+				continue;
+			}
+			return; // Found a secound item
+		}
+	}
+
+	PlaySFX(ItemInvSnds[ItemCAnimTbl[player.HoldItem._iCurs]]);
+
+	int cn = CURSOR_HAND;
+	if (it == 0) {
+		Stash.stashList.push_back(player.HoldItem);
+		it = Stash.stashList.size();
+	} else {
+		int stashIndex = it - 1;
+		cn = SwapItem(Stash.stashList[stashIndex], player.HoldItem);
+		for (auto &row : Stash.stashGrids[Stash.page]) {
+			for (auto &itemId : row) {
+				if (itemId == it)
+					itemId = 0;
+				if (itemId == -it)
+					itemId = 0;
+			}
+		}
+	}
+
+	AddItemToStashGrid(firstSlot, it, itemSize);
+
+	if (cn == CURSOR_HAND && !IsHardwareCursor())
+		SetCursorPos(MousePosition + Displacement(cursSize / 2));
+	NewCursor(cn);
+}
+
+void CheckStashCut(Point cursorPosition, bool automaticMove, bool dropItem)
+{
+	auto &player = Players[MyPlayerId];
+
+	if (withdrawGoldFlag) {
+		withdrawGoldFlag = false;
+		withdrawGoldValue = 0;
+	}
+
+	Point slot = { -1, -1 };
+
+	for (int y = 0; y < 10; y++) {
+		for (int x = 0; x < 10; x++) {
+			Point testSlot = { x, y };
+			Rectangle cell {
+				GetPanelPosition(UiPanels::Stash, StashCellPosition(testSlot)),
+				{ InventorySlotSizeInPixels.width + 1, InventorySlotSizeInPixels.height + 1 }
+			};
+
+			// check which inventory rectangle the mouse is in, if any
+			if (cell.Contains(cursorPosition)) {
+				slot = testSlot;
+				break;
+			}
+		}
+	}
+
+	if (slot == Point { -1, -1 }) {
+		return;
+	}
+
+	Item &holdItem = player.HoldItem;
+	holdItem._itype = ItemType::None;
+
+	bool automaticallyMoved = false;
+	bool automaticallyEquipped = false;
+
+	int16_t ii = Stash.stashGrids[Stash.page][slot.x][slot.y];
+	if (ii != 0) {
+		int iv = abs(ii) - 1;
+
+		holdItem = Stash.stashList[iv - 1];

```

Comment:
```suggestion
		holdItem = Stash.stashList[iv];
```

already subtracted 1 on line 242

Reply:
Thanks that saved me some debugging :)